src/main/core/support/emulated_time.rs

```diff
@@ -5,74 +5,99 @@ Deals with instances of time in a Shadow simulation.
 /// simulation.
 // Internally represented as Duration since the Unix Epoch.
 #[derive(Copy, Clone, Eq, PartialEq, Debug, PartialOrd, Ord)]
-pub struct EmulatedTime(Duration);
-
-/// The  Unix epoch (00:00:00 UTC on 1 January 1970)
-pub const UNIX_EPOCH: EmulatedTime = EmulatedTime(Duration::from_secs(0));
+pub struct EmulatedTime(c::EmulatedTime);
 
 // Duplicated from the EMULATED_TIME_OFFSET macro in definitions.h.
 pub(super) const SIMULATION_START_SEC: u64 = 946684800;
 pub const EMUTIME_INVALID: c::EmulatedTime = u64::MAX;
 pub const EMUTIME_MAX: c::EmulatedTime = u64::MAX - 1;
 pub const EMUTIME_MIN: c::EmulatedTime = 0;
 
-/// The start time of the simulation - 00:00:00 UTC on 1 January, 2000.
-pub const SIMULATION_START: EmulatedTime = EmulatedTime(Duration::from_secs(SIMULATION_START_SEC));
-
 impl EmulatedTime {
+    /// The start time of the simulation - 00:00:00 UTC on 1 January, 2000.
+    pub const SIMULATION_START: Self =
+        Self(SIMULATION_START_SEC * simulation_time::SIMTIME_ONE_SECOND);
+    /// The  Unix epoch (00:00:00 UTC on 1 January 1970)
+    pub const UNIX_EPOCH: Self = Self(0);
+
+    pub const MAX: Self = Self(EMUTIME_MAX);
+
     /// Get the instance corresponding to `val` SimulationTime units since the Unix Epoch.
     pub fn from_c_emutime(val: c::EmulatedTime) -> Option<Self> {
         if val == EMUTIME_INVALID {
             None
+        } else if val > EMUTIME_MAX {
+            None
         } else {
-            Some(Self(Duration::from_nanos(
-                val * simulation_time::SIMTIME_ONE_NANOSECOND,
-            )))
+            Some(Self(val))
         }
     }
 
     /// Convert to number of SimulationTime units since the Unix Epoch.
     pub fn to_c_emutime(val: Option<Self>) -> c::EmulatedTime {
         match val {
-            Some(val) => SimulationTime::to_c_simtime(Some(SimulationTime::from(
-                val.duration_since(&UNIX_EPOCH),
-            ))),
+            Some(v) => v.0,
             None => EMUTIME_INVALID,
         }
     }
 
     /// Get the instant corresponding to `val` time units since the simulation began.
     pub fn from_abs_simtime(val: SimulationTime) -> Self {
-        SIMULATION_START + *val
+        Self::SIMULATION_START + val
     }
 
     /// Convert to the SimulationTime since the simulation began.
     pub fn to_abs_simtime(self) -> SimulationTime {
-        SimulationTime::from(self.duration_since(&SIMULATION_START))
+        SimulationTime::from(self.duration_since(&Self::SIMULATION_START))
     }
 
-    /// Returns the duration since `earlier`, or panics if `earlier` is after `self`.
-    pub fn duration_since(&self, earlier: &EmulatedTime) -> Duration {
+    /// Returns the duration since `earlier`, or panics if `earlier` is after `self`, or
+    /// if the difference can't be represented as SimulationTime.
+    pub fn duration_since(&self, earlier: &EmulatedTime) -> SimulationTime {
```

Comment:
It still seems useful to be able to get durations from two `SimulationTime`s or two `EmulatedTime`s, and to have a distinction between durations and instances in time. But maybe that can be improved in the future when we don't need to worry about C types.

Reply:
> and to have a distinction between durations and instances in time. 

Yeah, that's the idea - `SimulationTime` is always a duration (except for some "legacy" uses in the C code), and `EmulatedTime` is always an instant. It might make sense to rename them something like `SimulatedDuration` and `SimulatedInstant` respectively, but probably not without likewise renaming the C types, which we probably shouldn't do before getting rid of all the remaining instances of `SimulationTime` being used to represent an instant in the C code.

> It still seems useful to be able to get durations from two SimulationTimes or two EmulatedTimes,

We can get a duration between two EmulatedTimes by using `EmulatedTime::duration_since` and `From`/`Into`. e.g. `Duration d = myEmulatedTime.duration_since(EmulatedTime::UNIX_EPOCH).into()`